### PR TITLE
[RISCV] Use shNadd for scalable stack offsets

### PR DESCRIPTION
**Pull request by preames**
If we need to multiply VLENB by 2, 4, or 8 and add it to the stack pointer, we can do so with a shNadd instead of separate shift and add instructions.

### CODE REVIEW

**Review by topperc — APPROVED** [review]
LGTM

**Review by lukel97 — APPROVED** [review]
LGTM

**Review by frasercrmck — APPROVED** [review]

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp [hunk 1]
```diff
@@ -204,10 +204,21 @@ void RISCVRegisterInfo::adjustReg(MachineBasicBlock &MBB,
...
-      .addReg(SrcReg).addReg(ScratchReg, RegState::Kill)
-      .setMIFlag(Flag);
+
+    if (ScalableAdjOpc == RISCV::ADD && ST.hasStdExtZba() &&
+        (NumOfVReg == 2 || NumOfVReg == 4 || NumOfVReg == 8)) {
+      unsigned Opc = NumOfVReg == 2 ? RISCV::SH1ADD :
+        (NumOfVReg == 4 ? RISCV::SH2ADD : RISCV::SH3ADD);
+      BuildMI(MBB, II, DL, TII->get(Opc), DestReg)
+          .addReg(ScratchReg, RegState::Kill)
+          .addReg(SrcReg, getKillRegState(KillSrcReg))
```
Comment (frasercrmck) [inline llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: If we're adding `getKillRegState(KillSrcReg)` here, should we do it below in the `else` too, for consistency? It always looks `false` so maybe that's why it wasn't done before.

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: We do have it on the else case. It's on the add at line 219.

Comment (frasercrmck) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: I had to do a double-take, but I meant on `SrcReg`.

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: Oops, you're right.  I misread.  Will follow up with a separate commit.

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: Addressed in 5ae9ffbd18fd93edbbc8efebe140aeb24cd763c2.  Sorry for the confusion here, I misread my own code when looking at your comment.  I think it was a combination of the swapped operand order, and that the shXadd case below does use the KillSrcReg flag.  But still, sorry for the unhelpful and confusing initial response.

Comment (frasercrmck) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:214]: No worries, thanks for the follow-up.

**Pull request merged**